src/Microsoft.Health.TaskManagement/JobHosting.cs

```diff
@@ -0,0 +1,250 @@
+                    }
+                    catch (Exception completeEx)
+                    {
+                        _logger.LogError(completeEx, "Job {jobId} failed to complete.", jobInfo.Id);
+                    }
+
+                    return;
+                }
+
+                try
+                {
+                    jobInfo.Status = JobStatus.Completed;
+                    await _queueClient.CompleteJobAsync(jobInfo, true, CancellationToken.None);
+                    _logger.LogInformation("Job {jobId} completed.", jobInfo.Id);
+                }
+                catch (Exception completeEx)
+                {
+                    _logger.LogError(completeEx, "Job {jobId} failed to complete.", jobInfo.Id);
+                }
+            }
+            finally
+            {
+                _activeJobsNeedKeepAlive.Remove(jobInfo.Id, out _);
+            }
+        }
+
+        private async Task KeepAliveSingleJobAsync(JobInfo jobInfo, CancellationTokenSource jobCancellationToken)
+        {
+            try
+            {
+                bool shouldCancel = false;
+                try
+                {
+                    JobInfo keepAliveResult = await _queueClient.KeepAliveJobAsync(jobInfo, CancellationToken.None);
+                    shouldCancel |= keepAliveResult.CancelRequested;
+                }
+                catch (JobNotExistException notExistEx)
+                {
+                    _logger.LogError(notExistEx, "Job {jobId} not exist or runid not match.", jobInfo.Id);
+                    shouldCancel = true;
+                }
+
+                if (shouldCancel && !jobCancellationToken.IsCancellationRequested)
+                {
+                    jobCancellationToken.Cancel();
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to keep alive on job {jobId}", jobInfo.Id);
+            }
+        }
+
+        private async Task KeepAliveJobsAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Start to keep alive job message.");
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Task intervalDelayTask = Task.Delay(TimeSpan.FromSeconds(JobHeartbeatIntervalInSeconds), CancellationToken.None);
+                KeyValuePair<long, Func<Task>>[] activeJobRecords = _activeJobsNeedKeepAlive.ToArray();
+
+                foreach ((long jobId, Func<Task> keepAliveFunc) in activeJobRecords)
+                {
+                    try
+                    {
+                        await keepAliveFunc();
+                    }
+                    catch (Exception ex)
+                    {
+                        _logger.LogError(ex, "Failed to keep alive on job {jobId}", jobId);
+                    }
+                }
+
+                await intervalDelayTask;
+            }
+
+            _logger.LogInformation("Stop to keep alive job message.");
```

Comment:
>"Stop to keep alive job message." [](http://example.com/codeflow?start=35&length=33)

We should include job id.

Reply:
this log is for the main keep alive thread, it is not for specific job. I put job id in keep single job alive method.